C error in `or {}` block

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.12 3cf0c92, press to see full `v doctor` output</summary>

|V full version      |V 0.4.12 ebf629dc4124a6207b18028cf89dc6910d25eaf3.3cf0c92
|:-------------------|:-------------------
|OS                  |linux, Linux version 6.17.9-arch1-1 (linux@archlinux) (gcc (GCC) 15.2.1 20251112, GNU ld (GNU Binutils) 2.45.1) #1 SMP PREEMPT_DYNAMIC Mon, 24 Nov 2025 15:21:09 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |5GB/15.34GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2025-12-17 15:29:14
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /tmp
|                    |
|Git version         |git version 2.52.0
|V git status        |weekly.2025.50-65-g3cf0c929
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.2.1 20251112
|gcc version         |gcc (GCC) 15.2.1 20251112
|clang version       |clang version 21.1.6
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.42

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  or_block_c_error.v && or_block_c_error`
```v
fn foo(a int) !int {
	if a < 0 {
		return error('foo')
	}
	return a
}

fn bar(a int) !int {
	return foo(a) or {
		if a < 0 {
			-1
		} else {
			return error('bar')
		}
	}
}

fn main() {
	assert bar(0)! == 0
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/or_block_c_error.01KCPEWAJFMZRHA0T0TE3VTY2F.tmp.c:4928: error: cannot convert 'struct _result_int' to 'int'
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Maybe a V checker error? IDK. Definitely not a C error.

**Additional info**

`return -1` instead of just `-1` resolves the issue.
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -200,8 +200,13 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 	mut exit_label := ''
 	if needs_tmp_var {
 		exit_label = g.new_tmp_var()
-		mut styp := g.styp(node.typ)
-		if g.inside_if_option || node.typ.has_flag(.option) {
+		node_typ := if g.inside_or_block {
+			node.typ.clear_option_and_result()
+		} else {
+			node.typ
+		}
+		mut styp := g.styp(node_typ)
+		if (g.inside_if_option || node_typ.has_flag(.option)) && !g.inside_or_block {
 			raw_state = g.inside_if_option
 			if node.typ != ast.void_type {
 				g.last_if_option_type = node.typ
@@ -214,7 +219,7 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 			}
 			g.inside_if_option = true
 			styp = styp.replace('*', '_ptr')
-		} else if node.typ.has_flag(.result) {
+		} else if node_typ.has_flag(.result) && !g.inside_or_block {
 			raw_state = g.inside_if_result
 			defer(fn) {
 				g.inside_if_result = raw_state
@@ -222,7 +227,7 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 			g.inside_if_result = true
 			styp = styp.replace('*', '_ptr')
 		} else {
-			g.last_if_option_type = node.typ
+			g.last_if_option_type = node_typ
 			defer(fn) {
 				g.last_if_option_type = tmp_if_option_type
 			}

--- a/vlib/v/tests/if_expr_in_or_block_test.v
+++ b/vlib/v/tests/if_expr_in_or_block_test.v
@@ -1,0 +1,105 @@
+fn foo(a int) !int {
+	if a < 0 {
+		return error('foo')
+	}
+	return a
+}
+
+fn bar(a int) !int {
+	return foo(a) or {
+		if a < 0 {
+			-1
+		} else {
+			return error('bar')
+		}
+	}
+}
+
+fn baz(a int) !int {
+	return foo(a) or {
+		if a < 0 {
+			-1
+		} else {
+			-2
+		}
+	}
+}
+
+fn qux(a int) ?int {
+	if a < 0 {
+		return none
+	}
+	return a
+}
+
+fn quux(a int) ?int {
+	return qux(a) or {
+		if a < 0 {
+			-1
+		} else {
+			return none
+		}
+	}
+}
+
+fn test_if_expr_in_result_or_block_with_return() {
+	assert bar(0)! == 0
+	assert bar(-1)! == -1
+	assert bar(1)! == 1
+}
+
+fn test_if_expr_in_result_or_block_simple() {
+	assert baz(0)! == 0
+	assert baz(-1)! == -1
+	assert baz(1)! == 1
+}
+
+fn test_if_expr_in_option_or_block_with_return() {
+	assert quux(0)? == 0
+	assert quux(-1)? == -1
+	assert quux(1)? == 1
+}
+
+fn test_nested_if_in_or_block() {
+	get_val := fn (x int) !int {
+		if x < 0 {
+			return error('negative')
+		}
+		return x * 2
+	}
+	result := get_val(-5) or {
+		if true {
+			if false {
+				100
+			} else {
+				-10
+			}
+		} else {
+			200
+		}
+	}
+	assert result == -10
+}
+
+fn test_if_expr_with_multiple_branches_in_or_block() {
+	compute := fn (x int) !int {
+		if x == 0 {
+			return error('zero')
+		}
+		return x
+	}
+	val := compute(0) or {
+		match err.msg() {
+			'zero' {
+				1
+			}
+			'negative' {
+				2
+			}
+			else {
+				3
+			}
+		}
+	}
+	assert val == 1
+}

